core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -722,9 +722,14 @@ private void updateStageExecutions()
                 PlanFragmentId fragmentId = subPlan.getFragment().getId();
                 StageId stageId = getStageId(fragmentId);
                 currentPlanStages.add(stageId);
-                if (isReadyForExecution(subPlan) && !stageExecutions.containsKey(stageId)) {
+                StageExecution stageExecution = stageExecutions.get(stageId);
+                if (isReadyForExecution(subPlan) && stageExecution == null) {
                     createStageExecution(subPlan, fragmentId.equals(rootFragmentId), nextSchedulingPriority++);
                 }
+                if (stageExecution != null && stageExecution.getState().equals(StageState.FINISHED) && !stageExecution.isClosed()) {
```

Comment:
nit: Do we need to check `!stageExecution.isClosed()` here?

Reply:
Yes, it's a pruning, so we don't call `abort` on the same stage for multiple times.

Reply:
This is what confuses me a little. You are checking the parent stage, but closing child stages. So you still may end up closing child stages several times until the parent stage itself is closed. Given the close is cheap (as you are checking if it's already closed) i wonder if it is worth extra complexity? (as it doesn't protect from close being called more than once but may create such an impression)

Reply:
I'd say `closeSourceExchanges` is not super cheap as it involves a loop which involves `getStageId` as well as a hashset lookup. Ideally we want to prune if possible.